shell/apple/emulator-osx/emulator-osx/osx-main.mm

```diff
@@ -131,14 +131,25 @@ bool emu_fast_forward()
     return settings.input.fastForwardMode;
 }
 
-int emu_single_frame(int w, int h)
+bool emu_vsync_enabled()
 {
-    if (!emu_frame_pending())
-        return 0;
+    return config::VSync;
+}
 
+bool emu_single_frame(int w, int h)
+{
     screen_width = w;
     screen_height = h;
-    return (int)mainui_rend_frame();
+    
+    //For DelayFrameSwapping: use while loop to call multple mainui_rend_frame() until rend_swap_frame(u32 fb_r_sof1)
+    while (mainui_enabled)
+    {
+        if (mainui_rend_frame())
+        {
+            return true;
+        }
+    }
+    return false;
```

Comment:
There's a risk of infinite loop here if the game isn't rendering. The only way out would be to exit the application.

Perhaps limit to 5 iterations max: 4 renders and the final swap.

Reply:
Right, although the `rend_framePending()` check guarded most of the infinite loop case, limiting the while loop also can solve the beachballing issue when the game isn't rendering also